presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -177,6 +177,8 @@ private PlanNodeWithCost getJoinNodeWithCost(Context context, JoinNode possibleJ
          *   always goes to the right), determining JOIN type is not that simple. As when
          *   choosing REPLICATED over REPARTITIONED join the cost of exchanging and building
          *   the hash table scales with the number of nodes where the build side is replicated.
+         *
+         *   TODO Decision about the distribution should be based on LocalCostEstimate only when PlanCostEstimate cannot be calculated. Otherwise cost comparator cannot take query.max-memory into account.
```

Comment:
But you would need to know how plan looks like downstream to actually take into account `query.max-memory` here. That's probably not feasible and heuristics here is fine

Reply:
Yes, we cannot make a fully educated decision, because we don't seePlanNodes higher up the plan tree. However, we can still make the best decision possible based on the information we have.

Reply:
Sure, but you cannot do comparisons against `query-max-memory-per-node` then as nodes downstream can cause mem to breach the limit